### PR TITLE
Add type annotations for a few torch.nn.modules

### LINKED ISSUE #46012: Enable torch.nn.modules.* typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.nn.modules.(loss/padding/pooling/rnn)]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #46012

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -210,6 +210,7 @@ def __init__(self, weight: Optional[Tensor] = None, size_average=None, ignore_in
         self.ignore_index = ignore_index
 
     def forward(self, input: Tensor, target: Tensor) -> Tensor:
+        assert self.weight is None or isinstance(self.weight, Tensor)
```
Comment (zou3519) [inline torch/nn/modules/loss.py:213]: Why do we need the asserts?

File: torch/nn/modules/padding.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ def __init__(self, value: float) -> None:
         self.value = value
 
     def forward(self, input: Tensor) -> Tensor:
-        return F.pad(input, self.padding, 'constant', self.value)
+        return F.pad(input, self.padding, 'constant', self.value)  # type: ignore[arg-type]
```
Comment (zou3519) [inline torch/nn/modules/padding.py:21]: Why is the `ignore` necessary here?

File: torch/nn/modules/rnn.py [hunk 3]
```diff
@@ -206,7 +205,7 @@ def permute_hidden(self, hx: Tensor, permutation: Optional[Tensor]):
             return hx
         return apply_permutation(hx, permutation)
 
-    def forward(self, input: Tensor, hx: Optional[Tensor] = None) -> Tuple[Tensor, Tensor]:
+    def forward(self, input: Tensor, hx: Optional[Tensor] = None) -> Tuple[Union[Tensor, PackedSequence], Tensor]:
```
Comment (zou3519) [inline torch/nn/modules/rnn.py:208]: `input` can also be a PackedSequence, right?

File: torch/nn/modules/rnn.py [hunk 4]
```diff
@@ -535,12 +537,12 @@ def check_forward_args(self, input: Tensor, hidden: Tuple[Tensor, Tensor], batch
         self.check_hidden_size(hidden[1], expected_hidden_size,
                                'Expected hidden[1] size {}, got {}')
 
-    def permute_hidden(self, hx: Tuple[Tensor, Tensor], permutation: Optional[Tensor]) -> Tuple[Tensor, Tensor]:
+    def permute_hidden(self, hx: Tuple[Tensor, Tensor], permutation: Optional[Tensor]) -> Tuple[Tensor, Tensor]:  # type: ignore
```
Comment (zou3519) [inline torch/nn/modules/rnn.py:540]: Why the ignore here?

File: torch/nn/modules/rnn.py [hunk 5]
```diff
@@ -526,7 +528,7 @@ class LSTM(RNNBase):
     def __init__(self, *args, **kwargs):
         super(LSTM, self).__init__('LSTM', *args, **kwargs)
 
-    def check_forward_args(self, input: Tensor, hidden: Tuple[Tensor, Tensor], batch_sizes: Optional[Tensor]):
+    def check_forward_args(self, input: Tensor, hidden: Tuple[Tensor, Tensor], batch_sizes: Optional[Tensor]):  # type: ignore
```
Comment (zou3519) [inline torch/nn/modules/rnn.py:531]: Why the ignore here?

File: torch/nn/modules/rnn.py [hunk 6]
```diff
@@ -283,7 +285,7 @@ def __setstate__(self, d):
 
     @property
     def all_weights(self) -> List[Parameter]:
-        return [[getattr(self, weight) for weight in weights] for weights in self._all_weights]
+        return [[getattr(self, weight) for weight in weights] for weights in self._all_weights]  # type: ignore
```
Comment (zou3519) [inline torch/nn/modules/rnn.py:288]: Is the problem that we're returning a List[List[Parameter]] and the signature is List[Parameter]?

**Review by zou3519 — CHANGES_REQUESTED** [review]
It looks like some TorchScript tests are failing. TorchScript makes use of the type annotation information

File: torch/nn/modules/padding.py (hunk 2 above)
Comment (guilhermeleobas) [reply torch/nn/modules/padding.py:21]: mypy infers the type of `self.padding` as `Union[Tensor, Module]` but `F.pad` expects padding as a `Sequence[int]`.

```
torch/nn/modules/padding.py:21: error: Argument 2 to "pad" has incompatible type "Union[Tensor, Module]"; expected "Sequence[int]"  [arg-type]
```

Replacing `# type: ignore[...]` by an assertion also crashes pytorch. I think I will defer annotating this module for now since it requires more work

File: torch/nn/modules/rnn.py (hunk 3 above)
Comment (guilhermeleobas) [reply torch/nn/modules/rnn.py:208]: Yes, you're right

File: torch/nn/modules/rnn.py (hunk 6 above)
Comment (guilhermeleobas) [reply torch/nn/modules/rnn.py:288]: Yes. `weight` here has type `Parameter?

File: torch/nn/modules/rnn.py (hunk 5 above)
Comment (guilhermeleobas) [reply torch/nn/modules/rnn.py:531]:
```
torch/nn/modules/rnn.py:531: error: Argument 2 of "check_forward_args" is incompatible with supertype "RNNBase"; supertype defines the argument type as "Tensor"  [override]
torch/nn/modules/rnn.py:531: note: This violates the Liskov substitution principle
torch/nn/modules/rnn.py:531: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

File: torch/nn/modules/rnn.py (hunk 4 above)
Comment (guilhermeleobas) [reply torch/nn/modules/rnn.py:540]:
```
torch/nn/modules/rnn.py:540: error: Argument 1 of "permute_hidden" is incompatible with supertype "RNNBase"; supertype defines the argument type as "Tensor"  [override]
torch/nn/modules/rnn.py:540: note: This violates the Liskov substitution principle
torch/nn/modules/rnn.py:540: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

File: torch/nn/modules/rnn.py [hunk 7]
```diff
@@ -535,12 +537,12 @@ def check_forward_args(self, input: Tensor, hidden: Tuple[Tensor, Tensor], batch
...
                                'Expected hidden[1] size {}, got {}')
 
-    def permute_hidden(self, hx: Tuple[Tensor, Tensor], permutation: Optional[Tensor]) -> Tuple[Tensor, Tensor]:
+    def permute_hidden(self, hx: Tuple[Tensor, Tensor], permutation: Optional[Tensor]) -> Tuple[Tensor, Tensor]:  # type: ignore
         if permutation is None:
             return hx
         return apply_permutation(hx[0], permutation), apply_permutation(hx[1], permutation)
 
-    @overload
+    @overload  # type: ignore
```
Comment (guilhermeleobas) [inline torch/nn/modules/rnn.py:545]: Same thing here:
```
torch/nn/modules/rnn.py:545: error: Signature of "forward" incompatible with supertype "RNNBase"  [override]
```

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (guilhermeleobas) [reply torch/nn/modules/loss.py:213]:
```
torch/nn/modules/loss.py:214: error: Argument "weight" to "nll_loss" has incompatible type "Union[Tensor, Module]"; expected "Optional[Tensor]"  [arg-type]
```

**Comment by guilhermeleobas** [comment]
Hi @zou3519, thanks for the review. I've marked the PR as draft and will work/experiment a bit more

File: torch/nn/modules/rnn.py (hunk 4 above)
Comment (rgommers) [reply torch/nn/modules/rnn.py:543]: This and the `@overload` on `forward` is the same issue as for `Module.forward`. That was quite tricky to resolve, but I believe the current solution works so far (see `_forward_unimplemented` and the comment above it in `nn/modules/module.py`).

I think that's best left to a separate PR, but it would be good to link to that solution, to save the next person working on this a lot of trouble. Can you add a comment here, and refer to it from the `@overload`'s futher down @guilhermeleobas?

**Review by rgommers — COMMENTED** [review]
Thanks @guilhermeleobas, this LGTM modulo adding a comment for the tricky subclassing issue.

**Comment by guilhermeleobas** [comment]
Thanks for the review, @rgommers

**Review by rgommers — APPROVED** [review]
LGTM now, and CI failures are all unrelated (connectivity/merge issues).
